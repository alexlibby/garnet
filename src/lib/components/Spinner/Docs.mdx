import { Canvas, Meta } from "@storybook/blocks";
import { ArgTypes } from "@storybook/blocks";
import * as SpinnerStories from "./Spinner.stories";

<Meta of={SpinnerStories} />

# Spinner

Spinner is the primary component. It has 2 possible states.

- [Circle](#circle)
- [Jumper](#jumper)

## Circle

This is the default state for our Spinner component:

<Canvas of={SpinnerStories.Circle} />

## Jumper

This is the Spinner component set as a Jumper:

<Canvas of={SpinnerStories.Jumper} />

## Properties of component

Below is a list of arguments available for this component:

<ArgTypes of={SpinnerStories} />
