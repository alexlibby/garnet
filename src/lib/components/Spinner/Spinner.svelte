<svelte:options customElement="garnet-spinner" />

<script>
  export let color = "#733635";
  export let duration = "0.75s";
  export let size = "60";
  export let variant = "";
  export let unit = 'px';
  export let pause = false;
  
  let durationUnit = duration.match(/[a-zA-Z]/)?.[0] ?? 's';
  let durationNum = duration.replace(/[a-zA-Z]/, '');
  
  const range = (size, startAt = 0) =>
  [...Array(size).keys()].map((i) => i + startAt);
</script>

<div class="garnet-spinner">
  <!-- Circle spinner -->
  {#if variant == "circle"}
    <div>
      <div
        class="circle"
        style="--size: {size}px; --color: {color}; --duration: {duration}"
      />
    </div>
  {/if}

  <!-- Jumper spinner -->
  {#if variant == "jumper"}
    <div style="--size: {size}px; --color: {color}; --duration: {duration};">
      {#each range(3, 1) as version}
        <div
          class="jumper"
          style="animation-delay: {(1 / 3) * (version - 1)}s;"
        />
      {/each}
    </div>
  {/if}

  <!-- Jellyfish spinner -->
  {#if variant == "jellyfish"}
  <div style="--size: 60px; --color: {color}; --motionOne: -12px; --motionTwo: 15px; --motionThree: -12px; --duration: 2.5s;">
      {#each range(6, 0) as version}
        <div
          class="ring"
          class:pause-animation={pause}
          style="animation-delay: {version * (+durationNum / 25)}{durationUnit}; width: {version *
            (+size / 6) +
            unit}; height: {(version * (+size / 6)) / 2 + unit}; "
        />
      {/each}
    </div>
  {/if}  
</div>

<style>
  .garnet-spinner {
    display: flex;
  }

  /* circle spinner */
  .circle {
    height: var(--size);
    width: var(--size);
    border-color: var(--color) transparent var(--color) var(--color);
    border-width: calc(var(--size) / 12);
    border-style: solid;
    border-radius: 50%;
    animation: var(--duration) linear 0s infinite normal none running
    rotateCircle;
  }
  
  @keyframes rotateCircle {
    0% {
      transform: rotate(0);
    }
    100% {
      transform: rotate(360deg);
    }
  }

  /* jumper spinner */
  .jumper {
    height: var(--size);
    width: var(--size);
    border-radius: 100%;
    animation-fill-mode: both;
    position: absolute;
    opacity: 0;
    background-color: var(--color);
    animation: bounce var(--duration) linear infinite;
  }
  
  @keyframes bounce {
    0% {
      opacity: 0;
      transform: scale(0);
    }
    5% {
      opacity: 1;
    }
    100% {
      opacity: 0;
      transform: scale(1);
    }
  }

  /* jellyfish spinner */
  .ring {
    position: absolute;
    border: 2px solid var(--color);
    border-radius: 50%;
    background-color: transparent;
    animation: motion var(--duration) ease infinite;
  }
  
  .pause-animation {
    animation-play-state: paused;
  }

  @keyframes motion {
    0% {
      transform: translateY(var(--motionOne));
    }
    50% {
      transform: translateY(var(--motionTwo));
    }
    100% {
      transform: translateY(var(--motionThree));
    }
  }
</style>
