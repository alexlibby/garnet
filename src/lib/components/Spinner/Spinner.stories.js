import Spinner from "./Spinner.svelte";
import { BADGE } from "storybook-addon-badges";

const BADGES = {
  ...BADGE,
  ALPHA: "Alpha",
};

export default {
  title: "Garnet UI Library/Action Components/Spinner",
  component: Spinner,
  argTypes: {
    color: "#733635",
    duration: "0.75s",
    size: "40",
    variant: "circle",
  },
  parameters: {
    docs: {
      story: {
        height: "100px",
      },
    },
  },
};

export const Circle = {
  args: {
    color: "#733635",
    duration: "0.75s",
    size: "40",
    variant: "circle",
  },
};

Circle.parameters = {
  badges: [BADGES.ALPHA],
  docs: { source: { type: "code" } },
};

export const Jumper = {
  args: {
    color: "#733635",
    duration: "1s",
    size: "60",
    variant: "jumper",
  },
};

Jumper.parameters = {
  badges: [BADGES.ALPHA],
  docs: { source: { type: "code" } },
};

export const Jellyfish = {
  args: {
    color: "#733635",
    duration: "2.5s",
    size: "60",
    variant: "jellyfish",
  },
};

Jellyfish.parameters = {
  badges: [BADGES.ALPHA],
  docs: { source: { type: "code" } },
};
