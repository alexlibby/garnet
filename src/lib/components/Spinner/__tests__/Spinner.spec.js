import { render } from "@testing-library/svelte";
import { axe } from "vitest-axe";
import * as AxeMatchers from "vitest-axe/matchers";
import Spinner from "../../Spinner/Spinner.svelte";

const SpinnerHTML = `<div class="garnet s-9fyBjQp-X91z"><div class="s-9fyBjQp-X91z"><div class="circle s-9fyBjQp-X91z" style="--size: 40px; --color: #733635; --duration: 0.75s;"></div></div></div>`;
expect.extend(AxeMatchers);

describe("Tests for Spinner", () => {
  const props = {
    color: "#733635",
    duration: "0.75s",
    size: "40",
    variant: "circle",
  };

  it("should render properly", async () => {
    const result = render(Spinner, { props });
    expect(() => result).not.toThrow();
  });

  it("get a snapshot of component", () => {
    // Create a new container for the test and mount component
    const host = document.createElement("div");
    document.body.appendChild(host);
    const instance = new Spinner({ target: host, props: props });

    // check that all rendered OK
    expect(instance).toBeTruthy();

    // Take snapshot
    expect(host.innerHTML).toMatchSnapshot();
  });

  test("should demonstrate no issues with accessibility", async () => {
    const render = () => SpinnerHTML;
    const result = render();

    expect(await axe(result)).toHaveNoViolations();
  });
});
