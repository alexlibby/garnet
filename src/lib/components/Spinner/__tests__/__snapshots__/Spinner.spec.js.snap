// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for Spinner > get a snapshot of component 1`] = `"<div class="garnet svelte-umkm7n"><div><div class="circle svelte-umkm7n" style="--size: 40px; --color: #733635; --duration: 0.75s;"></div></div>  </div>"`;
